### ISSUE TITLE
gtk-doc build fails: configure: could not find DocBook XML DTD V4.3 in XML catalog

### ISSUE THREAD

**Issue by superNick459** [LLNL/spack#25420]
### Steps to reproduce the issue

`gtk-doc` fails during config stage using:

- spack@develop [f0875b2](https://github.com/spack/spack/commit/f0875b2fef8e495809854141785c1fc7f5d86407)

- Ubuntu 20.04, GCC 9.3.0
- Docker container: `ecpe4s/ubuntu20.04-runner-x86_64:2021-07-01`

Spec: [gtk-doc-spack.yaml.txt](https://github.com/spack/spack/files/6984545/gtk-doc_spack.yaml.txt)

How to reproduce:
```
$ git clone https://github.com/spack/spack
$ cd spack
$ git checkout f0875b2
$ . share/spack/setup-env.sh
$ spack env create test ~/gtk-doc-spack.yaml
$ spack env activate test
$ spack buildcache keys -it
$ spack concretize -f
$ spack install
==> Installing gtk-doc-1.32-2z5hjsgkzaemspqqxlgfaerb5wrdhean
==> No binary for gtk-doc-1.32-2z5hjsgkzaemspqqxlgfaerb5wrdhean found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/08/0890c1f00d4817279be51602e67c4805daf264092adc58f9c04338566e8225ba.tar.gz
==> Applied patch /root/spack/var/spack/repos/builtin/packages/gtk-doc/build.patch
==> gtk-doc: Executing phase: 'autoreconf'
==> gtk-doc: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/root/spack-stage/spack-stage-gtk-doc-1.32-2z5hjsgkzaemspqqxlgfaerb5wrdhean/spack-src/configure' '--prefix=/root/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/gtk-doc-1.32-2z5hjsgkzaemspqqxlgfaerb5wrdhean'

1 error found in build log:
     126    checking for dblatex... no
     127    checking for fop... no
     128    configure: WARNING: neither dblatex nor fop found, so no pdf output from xml
     129    checking for XML catalog...
     130    checking for xmlcatalog... /root/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/libxml2-2.9.10-qlt5ajqqmtin3atqcycnv67oex63cvny/bin/xmlcatalo
            g
     131    checking for DocBook XML DTD V4.3 in XML catalog... not found
  >> 132    configure: error: could not find DocBook XML DTD V4.3 in XML catalog

See build log for details:
  /tmp/root/spack-stage/spack-stage-gtk-doc-1.32-2z5hjsgkzaemspqqxlgfaerb5wrdhean/spack-build-out.txt
```

### Information on your system

```
* **Spack:** 0.16.2-3907-f0875b2fef
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-zen3
* **Concretizer:** original
```

### Additional information

Build log: [gtk-doc-build-out.txt](https://github.com/spack/spack/files/6984553/gtk-doc-build-out.txt)

Configure log: [spack-build-02-configure-out.txt](https://github.com/spack/spack/files/6984557/spack-build-02-configure-out.txt)

@lee218llnl @adamjstewart 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by lorddavidiii** [comment]
So, may this related, I recognized that after #24300 the name for the xml catalogs changes from `catalog.xml` to `catalog`, hence I had to reinstall  `docbook-xsl` and `docbook-xml` to fix this.

 I just checked, I install these to packages one commit before the PR, ie 970bf4318c, them `XML_CATALOG_FILES` points to `catalog.xml`, but after checking out th PR commit, ie 4fd8640586816d59693b264bcbd5e1378b0963ff, the export `XML_CATALOG_FILES` now points to the `catalog` file after loading, which does not exist, because the package was not rebuild and have only `catalog.xml`, but not a `catalog` file.

**Comment by glennpj** [comment]
The `XML_CATALOG_FILES` points to the file with the entities, so the name itself should not matter. The `gtk-doc package builds fine on a CentOS-7 system. That likely means that the system installed catalog is, or is not, coming into play, possibly due to the environment variable being ignored.

**Comment by glennpj** [comment]
It could also be something amiss with the entities in the docbook packages. I will look into that as well.

**Comment by glennpj** [comment]
@superNick459 Can you try PR #25569 and see if that fixes the problem.

**Comment by alalazo** [comment]
@superNick459 ?

**Comment by bernhardkaindl** [comment]
@alalazo @superNick459  I've tested if PR #25569 fixes this issue:

No, PR #25569 doesn't fix this issue, but it is fixed by the now-merged #22845. The commit message says:
>    **NOTE**: Because we have switched the spec format, this PR changes Spack's hashing algorithm.  This means that after this commit, Spack will think a lot of things need rebuilds.

And that's the effect. With this commit on develop now, the old precompiled docbook-xml package isn't fetched from https://cache.e4s.io but recompiled, with @glennpj new local docbook catalog filenames and because everything is built from source, the build succeeds.

I'll add a comment going deeper of how this would have been prevented and fixes without a change in the global hash algorithm in a review of #25569.

You can close this issue now, the bug is thereby obsolete.

**Comment by alalazo** [comment]
@bernhardkaindl @glennpj Ok, that seems reasonable. I'll close the issue, but feel free to reopen if you think there's anything else that needs to be discussed here.

**Comment by glennpj** [comment]
Basically, the fixes in #24300 were not getting picked up because there was no change in the hash. Although this is fixed now it seems like this type of thing could pop up for other packages going forward.

@alalazo Is there a way to force a hash update of a package if none of the changes from a PR otherwise trigger a hash change? It almost seems like the package.py itself would need to be part of the hash but that may be too simplistic.

**Comment by alalazo** [comment]
> It almost seems like the package.py itself would need to be part of the hash but that may be too simplistic.

@glennpj Right now we have 3 different kind of hashes. Details are more complicated but roughly:
1. DAG hash (accounts only for link and run deps from root)
2. Build hash (accounts for all deps from root)
3. Full hash (same as build hash + accounts for package.py)

I need to double check, but I think for buildcaches we are using the build hash to avoid excessive rebuilds at least until #25310 lands in develop. Once that is in the idea would be, in the long term, to eliminate the dag/build hash and converge to a unique concept that is used everywhere in Spack.
